### PR TITLE
JetStream Snapshots

### PR DESCRIPTION
**Pull request by derekcollison**
We now support snapshots and restores for JetStream streams.
Snapshots are gzipped tar files of the file store directory for a stream.
This tries to protect the server and stream data and not bloat the server.

/cc @nats-io/core

### CODE REVIEW

**Comment by derekcollison** [comment]
Check the restore over NATS messages. We send request which returns a response with a DeliverSubject (mirroring snapshots) and we send chunks to there.

Problem is we can only ask if the stream exists post sending + flush, not a clean way to communicate errors during.

Don't want to simply send to the open restore API endpoint streams of data.

One possible idea is when you send the nil in you can attach a reply and we can send back a response indicating success or failure potentially.

File: server/jetstream_api.go [hunk 1]
```diff
@@ -287,6 +301,35 @@ type JSApiMsgDeleteResponse struct {
 
 const JSApiMsgDeleteResponseType = "io.nats.jetstream.api.v1.stream_msg_delete_response"
 
+type JSApiStreamSnapshotRequest struct {
+	// Subject to deliver the chunks to for the snapshot.
+	DeliverSubject string `json:"deliver_subject"`
+	// Do not include consumers in the snapshot.
+	NoConsumers bool `json:"no_consumers,omitempty"`
+	// Optional chunk size preference. Otherwise server selects.
+	ChunkSize int `json:"chunk_size,omitempty"`
```
Comment (ripienaar) [inline server/jetstream_api.go:310]: I am still at 2 minds about the max time a snapshot might run, in many industries its common to say operations interruptions are strictly time bucketed and they would rather fix what makes backups slow than have backup overrun ime, but we can always add that later

File: server/jetstream_api.go [hunk 2]
```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
...
+	}
+
+	var req JSApiStreamSnapshotRequest
+	if err := json.Unmarshal(msg, &req); err != nil {
+		resp.Error = jsInvalidJSONErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !IsValidSubject(req.DeliverSubject) {
+		resp.Error = jsBadRequestErr
```
Comment (ripienaar) [inline server/jetstream_api.go:1172]: we really don't want any `jsBadRequestErr` anywhere :) Better to send a `&ApiError{Code: 400, Description: "invalid delivery subject"}` or similar

File: server/jetstream_api.go [hunk 3]
```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
...
+		case <-time.After(2 * time.Second):
+		}
+	}
+
+	// Create our ack flow handler.
+	// This is very simple for now.
+	acks := make(chan struct{}, 1)
+	acks <- struct{}{}
+
+	ackSubj := fmt.Sprintf("_snapshot_.%s", nuid.Next())
```
Comment (ripienaar) [inline server/jetstream_api.go:1224]: can we make this `$JS.something` so its clearly associated with the lifecycle and assists with ACLs, acks are `$JS.ACK.>` so maybe `$JS.SNAPSHOT.ACK.>` to give us some space to expand into there for other flow control messages in future - things like client notifying you about windowing etc

File: server/jetstream_api.go [hunk 4]
```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
...
+		if len(chunk)+n > chunkSize {
+			// Wait on acks for flow control.
+			// Wait up to 10ms for now if none received.
+			select {
+			case <-acks:
+			case <-inch: // Lost interest
+				goto done
+			case <-time.After(10 * time.Millisecond):
+			}
+			mset.sendq <- &jsPubMsg{reply, _EMPTY_, ackSubj, chunk, nil, 0}
```
Comment (ripienaar) [inline server/jetstream_api.go:1258]: this being the same sendq as normal message delivery should we be concerned about backups consuming all the senq buffer and starving ongoing work?

Especially outgoing acks from stream submits as those can be quite time sensitive and we don't want to recreate the issues TCP have when the there's a lot of pressure and acks come late causing many retransmissions.

File: server/stream.go [hunk 5]
```diff
@@ -915,3 +934,166 @@ func (mset *Stream) ackMsg(obs *Consumer, seq uint64) {
...
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of snapshot
+		}
+		if err != nil {
+			return nil, err
+		}
+		fpath := path.Join(sdir, filepath.Clean(hdr.Name))
+		pdir := filepath.Dir(fpath)
+		os.MkdirAll(pdir, 0755)
```
Comment (ripienaar) [inline server/stream.go:1004]: I know we use `755` for all the dirs now so its nothing new, but probably `750` be better?

File: server/jetstream_api.go (hunk 1 above)
Comment (ripienaar) [reply server/jetstream_api.go:310]: What do you think is a sane minimum for chunk size? I'll put it in the schema you dont need to do anything but like even 1kb seems just dumb as a chunk size so I'll give some guidance in the schemas

File: server/jetstream_api.go [hunk 6]
```diff
@@ -287,6 +301,35 @@ type JSApiMsgDeleteResponse struct {
...
+	BlkSize int `json:"blk_size"`
+}
+
+const JSApiStreamSnapshotResponseType = "io.nats.jetstream.api.v1.stream_snapshot_response"
+
+// JSApiStreamRestoreResponse is the direct response to the restore request.
+type JSApiStreamRestoreResponse struct {
+	ApiResponse
+	// Subject to deliver the chunks to for the snapshot restore.
+	DeliverSubject string `json:"deliver_subject"`
```
Comment (ripienaar) [inline server/jetstream_api.go:328]: is chunk size pretty much just up to the client to do whatever it likes?

File: server/jetstream_api.go [hunk 7]
```diff
@@ -287,6 +301,35 @@ type JSApiMsgDeleteResponse struct {
...
+	NoConsumers bool `json:"no_consumers,omitempty"`
+	// Optional chunk size preference. Otherwise server selects.
+	ChunkSize int `json:"chunk_size,omitempty"`
+}
+
+// JSApiStreamSnapshotResponse is the direct response to the snapshot request.
+type JSApiStreamSnapshotResponse struct {
+	ApiResponse
+	// Estimate of number of blocks for the messages.
+	NumBlks int `json:"num_blks"`
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/jetstream_api.go:317]: Can't wait to see your progress bar - lol

File: server/jetstream_api.go [hunk 8]
```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
...
+	acc := c.acc
+
+	var resp = JSApiStreamSnapshotResponse{ApiResponse: ApiResponse{Type: JSApiStreamSnapshotResponseType}}
+	if !acc.JetStreamEnabled() {
+		resp.Error = jsNotEnabledErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if isEmptyRequest(msg) {
+		resp.Error = jsBadRequestErr
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/jetstream_api.go:1153]: Possibly, still technically a bad request.

File: server/jetstream_api.go (hunk 1 above)
Comment (derekcollison) [reply server/jetstream_api.go:310]: Normally should not be set at all. We could event remove it IMO. Server picks 64k atm.

File: server/jetstream_api.go (hunk 2 above)
Comment (derekcollison) [reply server/jetstream_api.go:1172]: ok

File: server/jetstream_api.go (hunk 3 above)
Comment (derekcollison) [reply server/jetstream_api.go:1224]: This would be an allow_response dynamic permission.

Comment (derekcollison) [reply server/jetstream_api.go:1224]: Not a fixed one IMO.

File: server/jetstream_api.go (hunk 4 above)
Comment (derekcollison) [reply server/jetstream_api.go:1258]: The snapshots are pretty conservative atm. We could move them odd though if it becomes a problem, will add a comment.

File: server/stream.go (hunk 5 above)
Comment (derekcollison) [reply server/stream.go:1004]: ok will update.

File: server/jetstream_api.go (hunk 6 above)
Comment (derekcollison) [reply server/jetstream_api.go:328]: Yes they can pretty much pick anything. We could send back a suggested chunk size but not sure it matters. We can add one in future if needed.

File: server/jetstream_api.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ripienaar) [reply server/jetstream_api.go:1153]: yeah they are all bad - but its about user feedback :)

File: server/jetstream_api.go (hunk 3 above)
Comment (derekcollison) [reply server/jetstream_api.go]: After discussion agree we need better prefixes etc. Fixed.

File: server/jetstream_api.go (hunk 1 above)
Comment (derekcollison) [reply server/jetstream_api.go:314]: I updated to 128k default.

File: server/jetstream_api.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/jetstream_api.go:1157]: We can do a round of updates in separate PR.

**Comment by derekcollison** [comment]
Yep.. ;)

File: server/jetstream_api.go [hunk 9]
```diff
@@ -1033,6 +1084,195 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
...
+			tfile.Seek(0, 0)
+			// TODO(dlc) - no way right now to communicate back.
+			acc.RestoreStream(stream, tfile)
+			tfile.Close()
+			os.Remove(tfile.Name())
+			c.processUnsub(sub.sid)
+			return
+		}
+		// Append chunk to temp file.
+		tfile.Write(msg)
```
Comment (ripienaar) [inline server/jetstream_api.go:1136]: need to catch errors here @derekcollison

Comment (ripienaar) [reply server/jetstream_api.go:1136]: also the lack of ack here is a bit of a problem, there's no way for a restore to know its happening/working until after its potentially sent gigs of data and then seeing the stream doesnt exist, so we need to do something better here i think

**Pull request merged**